### PR TITLE
Inline action of Like Post

### PR DESCRIPTION
**Pull request by jarvislin**
Part of #20024 

This PR contains:
1. Implement the inline action of Like Post
2. Refactors (mostly in NoteAdapter)
-----

## To Test:

1. Open the Jetpack app and switch to the notifications tab.
2. Find a post and click on the star icon in the notification item.
3. The Like status should be updated.
4. Done, thank you!

-----

## Regression Notes

1. Potential unintended areas of impact

    - Notification tab

3. What I did to test those areas of impact (or what existing automated tests I relied on)

    - Manual

4. What automated tests I added (or what prevented me from doing so)

    - None

-----

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## Testing Checklist:

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [x] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WordPress/src/main/java/org/wordpress/android/ui/reader/actions/ReaderPostActions.java [hunk 1]
```diff
@@ -82,43 +83,52 @@ public static boolean performLikeActionLocal(final ReaderPost post,
...
                                                final boolean isAskingToLike,
                                                final long wpComUserId,
-                                               final ActionListener actionListener) {
+                                               @Nullable final ActionListener actionListener) {
+        if (post != null) {
+            performLikeActionRemote(post, post.blogId, post.postId, isAskingToLike, wpComUserId, actionListener);
+        }
+    }
+
+    public static void performLikeActionRemote(
```
Comment (jarvislin) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/actions/ReaderPostActions.java:95]: Overload the `performLikeActionRemote()` function to handle the nullalbe `ReaderPost`

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderLikeTable.java [hunk 2]
```diff
@@ -77,14 +77,18 @@ public static void setCurrentUserLikesPost(ReaderPost post, boolean isLiked, lon
         if (post == null) {
             return;
         }
+        setCurrentUserLikesPost(post.postId, post.blogId, isLiked, wpComUserId);
+    }
+
+    public static void setCurrentUserLikesPost(long postId, long blogId, boolean isLiked, long wpComUserId) {
```
Comment (jarvislin) [inline WordPress/src/main/java/org/wordpress/android/datasets/ReaderLikeTable.java:83]: Overload the `setCurrentUserLikesPost()` function to handle the nullalbe `ReaderPost`

File: WordPress/src/main/java/org/wordpress/android/models/Note.java [hunk 3]
```diff
@@ -390,13 +389,49 @@ private JSONObject getCommentActions() {
         return mActions;
     }
 
-    /*
+    @NonNull private JSONObject getPostActions() {
```
Comment (jarvislin) [inline WordPress/src/main/java/org/wordpress/android/models/Note.java:392]: This is copied from `getCommentActions()`. It looks really bad, I plan to use Gson to refactor this part when we move Note to fluxc project.

File: WordPress/src/main/java/org/wordpress/android/ui/notifications/adapters/NotesAdapter.kt [hunk 4]
```diff
@@ -57,7 +56,6 @@ class NotesAdapter(
     private val textIndentSize: Int
     private val dataLoadedListener: DataLoadedListener
     private val onLoadMoreListener: OnLoadMoreListener?
-    private val notes = ArrayList<Note>()
```
Comment (jarvislin) [inline WordPress/src/main/java/org/wordpress/android/ui/notifications/adapters/NotesAdapter.kt:60]: We don't need two `ArrayList<Note>` so let's remove one of them for better performance

File: WordPress/src/main/java/org/wordpress/android/ui/notifications/adapters/NotesAdapter.kt [hunk 5]
```diff
@@ -101,44 +99,30 @@ class NotesAdapter(
...
-        return NoteViewHolder(view)
-    }
+            .let { NoteViewHolder(it) }
 
-    private fun getNoteAtPosition(position: Int): Note? {
-        return if (isValidPosition(position)) {
-            filteredNotes[position]
-        } else null
-    }
+    private fun getNoteAtPosition(position: Int): Note? = if (isValidPosition(position)) {
```
Comment (jarvislin) [inline WordPress/src/main/java/org/wordpress/android/ui/notifications/adapters/NotesAdapter.kt:118]: > Prefer using an expression body for functions with the body consisting of a single expression.
https://kotlinlang.org/docs/coding-conventions.html#functions

File: WordPress/src/main/java/org/wordpress/android/ui/notifications/adapters/NotesAdapter.kt [hunk 6]
```diff
@@ -450,37 +437,18 @@ class NotesAdapter(
...
-            filteredNotes: ArrayList<Note>,
-            notes: ArrayList<Note>,
+            notes: List<Note>,
             filter: FILTERS
-        ) {
-            filteredNotes.clear()
-            if (notes.isEmpty() || filter == FILTERS.FILTER_ALL) {
-                filteredNotes.addAll(notes)
-                return
+        ): ArrayList<Note> = notes.filter { note ->
```
Comment (jarvislin) [inline WordPress/src/main/java/org/wordpress/android/ui/notifications/adapters/NotesAdapter.kt:444]: Make it more "Kotlin" and shorter

File: WordPress/src/main/java/org/wordpress/android/models/Note.java [hunk 7]
```diff
@@ -390,13 +390,51 @@ private JSONObject getCommentActions() {
         return mActions;
     }
 
-    /*
+    @NonNull
+    private JSONObject getPostActions() {
```
Comment (jarvislin) [inline WordPress/src/main/java/org/wordpress/android/models/Note.java:394]: This is copied from getCommentActions(). It looks really bad, I plan to use Gson to refactor this part when we move Note to fluxc project.

**Review by antonis — APPROVED** [review]
Awesome work @jarvislin 🥇 
I've tested the implementation on a Pixel 5/Android 14 and everything worked as described. The code also looks good 🎉 
Thank you for also improving the code along the way 🙇

**Pull request merged**